core/state/state_object.go

```diff
@@ -217,6 +217,18 @@ func (self *stateObject) setState(key, value common.Hash) {
 // updateTrie writes cached storage modifications into the object's storage trie.
 func (self *stateObject) updateTrie(db Database) Trie {
 	tr := self.getTrie(db)
+
+	// save the originStorage first, ensure all non empty keys in originStorage saved to trie.
+	// if the object is copyed from another one,
+	// the dirtyStorage is empty and the originStorage is not empty.
+	for key, value := range self.originStorage {
+		if value != (common.Hash{}) {
+			// Encoding []byte cannot fail, ok to ignore the error.
+			v, _ := rlp.EncodeToBytes(bytes.TrimLeft(value[:], "\x00"))
+			self.setError(tr.TryUpdate(key[:], v))
+		}
+	}
```

Comment:
Could you clarify why there is a need to save `originStorage` to the trie? Ostensibly, they already reflect what's in the trie?

Reply:
IMO, we have two kinds of `preimage` provider, one is the account addresses, the other one is the contract account's storage keys. But the later one's preimags were not saved into the  underlying trie, so next time if you want to get the storage keys back, you got nothing.  Such as you want to dump a specified block like `geth dump xxx`, you can only get all the accounts, but the storage in contract is empty, because the keys are missing.